File: vantare-v2/internal/license/plan.go
package license

import "sort"

// PlanLabel is the user-facing classification of a license. It is intentionally
// non-overlapping so the UI can render one and only one row.
type PlanLabel string

const (
	PlanFree         PlanLabel = "free"
	PlanPaidOverlays PlanLabel = "paid_overlays"
	PlanPaidEngineer PlanLabel = "paid_engineer"
	PlanSuite        PlanLabel = "suite"
	PlanUnknown      PlanLabel = "unknown"
)

// PlanStatus is the high-level gating state surfaced to the UI together with
// the plan label. It collapses the seven State values into six buckets the
// settings page renders as badges.
type PlanStatus string

const (
	PlanStatusActive       PlanStatus = "active"
	PlanStatusGrace        PlanStatus = "grace"
	PlanStatusBlocked      PlanStatus = "blocked"
	PlanStatusFree         PlanStatus = "free"
	PlanStatusAnonymous    PlanStatus = "anonymous"
	PlanStatusUnconfigured PlanStatus = "unconfigured"
)

// PlanSummary is the wire-shaped status consumed by AccountSettings.
type PlanSummary struct {
	Label  PlanLabel
	Status PlanStatus
}

// ClassifyPlan returns the user-facing plan label for a set of entitlements.
// The rules are intentionally simple and follow docs/stripe-integration-plan.md:
//   - bundle / beta_access / founder / pro_founder / visionary_backer  -> suite
//   - overlays (alone or with supporter)                              -> paid_overlays
//   - engineer (alone or with supporter)                              -> paid_engineer
//   - supporter alone                                                 -> paid_overlays (Supporter is documented as Overlays+badge)
//   - ac_lua_pack alone                                                -> free (add-on, does not unlock the app)
//   - empty                                                            -> free
//   - any other unrecognized token                                     -> unknown
func ClassifyPlan(entitlements []Entitlement) PlanLabel {
	// Deduplicate while preserving presence. Empty strings are ignored so the
	// caller can pass a slice with a stray placeholder without surprises.
	has := map[Entitlement]bool{}
	for _, e := range entitlements {
		if e == "" {
			continue
		}
		has[e] = true
	}
	if len(has) == 0 {
		return PlanFree
	}

	// Suite first: any tier that maps to Overlays + Engineer supersedes
	// standalone entitlements. Overlays+Engineer together also count as the
	// Suite experience even without an explicit bundle token, mirroring the
	// paywall "Si tienes Overlays + Engineer, ya tienes Suite" rule.
	switch {
	case has[EntitlementBundle],
		has[EntitlementBetaAccess],
		has[EntitlementFounder],
		has[EntitlementProFounder],
		has[EntitlementVisionaryBacker],
		has[EntitlementOverlays] && has[EntitlementEngineer]:
		return PlanSuite
	}

	if has[EntitlementOverlays] || has[EntitlementSupporter] {
		return PlanPaidOverlays
	}
	if has[EntitlementEngineer] {
		return PlanPaidEngineer
	}
	if has[EntitlementACLuaPack] {
		return PlanFree
	}

	// Unknown entitlement keys fall back to "unknown" rather than "free" so
	// the UI can flag stale or misconfigured data instead of silently granting
	// access.
	return PlanUnknown
}

// ClassifyStatus collapses the license State into the PlanStatus buckets the
// settings page renders. Anonymous users are kept distinct so the UI can show
// "Sin sesión" instead of "Free" when there is no logged-in user.
func ClassifyStatus(state State) PlanStatus {
	switch state {
	case StateActive:
		return PlanStatusActive
	case StateGrace:
		return PlanStatusGrace
	case StateExpired, StateDeviceLimit:
		return PlanStatusBlocked
	case StateAuthenticatedNoEntitlement:
		return PlanStatusFree
	case StateAnonymous:
		return PlanStatusAnonymous
	case StateUnconfigured:
		return PlanStatusUnconfigured
	default:
		return PlanStatusFree
	}
}

// BuildSummary produces the summary the AccountSettings card renders.
func BuildSummary(state State, entitlements []Entitlement) PlanSummary {
	status := ClassifyStatus(state)
	label := ClassifyPlan(entitlements)

	// Blocked and anonymous states override the label so the UI never shows
	// "Suite" while the user is logged out or the subscription is dead.
	// Unconfigured is a configuration error, not a block: it keeps the label
	// so the UI can show an actionable message without a false paywall.
	if status == PlanStatusBlocked || status == PlanStatusAnonymous {
		return PlanSummary{Label: label, Status: status}
	}

	// Active and grace states only keep a suite/paid label when there is at
	// least one entitlement. Otherwise we surface "free".
	if status == PlanStatusActive || status == PlanStatusGrace {
		if label == PlanUnknown {
			return PlanSummary{Label: PlanFree, Status: status}
		}
		return PlanSummary{Label: label, Status: status}
	}

	return PlanSummary{Label: PlanFree, Status: status}
}

// SortedEntitlements returns a sorted copy of entitlements so the UI can render
// a stable list without changing call sites.
func SortedEntitlements(entitlements []Entitlement) []Entitlement {
	out := make([]Entitlement, len(entitlements))
	copy(out, entitlements)
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

File: vantare-v2/frontend/src/lib/plan.ts
import type { Entitlement, LicenseState } from "./license-types";

// PlanLabel mirrors internal/license.PlanLabel.
export type PlanLabel =
  | "free"
  | "paid_overlays"
  | "paid_engineer"
  | "suite"
  | "unknown";

// PlanStatus mirrors internal/license.PlanStatus.
export type PlanStatus =
  | "active"
  | "grace"
  | "blocked"
  | "free"
  | "anonymous"
  | "unconfigured";

export type PlanSummary = {
  label: PlanLabel;
  status: PlanStatus;
};

// The canonical labels shown in the AccountSettings card and the PaywallScreen
// banner. Kept in Spanish to match the rest of the beta UI.
export const PLAN_LABELS: Record<PlanLabel, string> = {
  free: "Free",
  paid_overlays: "Overlays",
  paid_engineer: "Engineer",
  suite: "Suite",
  unknown: "Sin clasificar",
};

export const PLAN_STATUS_LABELS: Record<PlanStatus, string> = {
  active: "Activo",
  grace: "Periodo de gracia",
  blocked: "Bloqueado",
  free: "Sin suscripción",
  anonymous: "Sin sesión",
  unconfigured: "Configuración incompleta",
};

function normaliseEntitlements(entitlements: Entitlement[] | null | undefined): Entitlement[] {
  return Array.isArray(entitlements) ? entitlements : [];
}

export function classifyPlan(entitlements: Entitlement[] | null | undefined): PlanLabel {
  const has = new Set<Entitlement>();
  for (const e of normaliseEntitlements(entitlements)) {
    if (!e) continue;
    has.add(e);
  }
  if (has.size === 0) return "free";

  if (
    has.has("bundle") ||
    has.has("beta_access") ||
    has.has("founder") ||
    has.has("pro_founder") ||
    has.has("visionary_backer") ||
    (has.has("overlays") && has.has("engineer"))
  ) {
    return "suite";
  }
  if (has.has("overlays") || has.has("supporter")) return "paid_overlays";
  if (has.has("engineer")) return "paid_engineer";
  if (has.has("ac_lua_pack")) return "free";

  return "unknown";
}

export function classifyStatus(state: LicenseState | null): PlanStatus {
  switch (state) {
    case "active":
      return "active";
    case "grace":
      return "grace";
    case "expired":
    case "device-limit":
      return "blocked";
    case "authenticated-no-entitlement":
      return "free";
    case "anonymous":
      return "anonymous";
    case "unconfigured":
      return "unconfigured";
    default:
      return "free";
  }
}

export function buildSummary(
  state: LicenseState | null,
  entitlements: Entitlement[] | null | undefined,
): PlanSummary {
  const status = classifyStatus(state);
  const label = classifyPlan(entitlements);

  if (status === "blocked" || status === "anonymous") {
    return { label, status };
  }
  // Unconfigured is a configuration error, not a block: keep the label so
  // the UI can show an actionable message without a false paywall.
  if (status === "unconfigured") {
    return { label, status };
  }
  if (status === "active" || status === "grace") {
    if (label === "unknown") return { label: "free", status };
    return { label, status };
  }
  return { label: "free", status };
}

export function sortedEntitlements(entitlements: Entitlement[] | null | undefined): Entitlement[] {
  return [...normaliseEntitlements(entitlements)].sort();
}
